### PR TITLE
Feature/photocathode model

### PR DESCRIPTION
**Pull request by kmtsui**
~~Prerequisite: Developed with geant4.10.05.p01 and root_v6.22.06, need to be tested with older (official) software versions.~~

Implement more realistic photocathode model to treat reflection/transmission/absorption of photons on the photocathode glass (OpGlassCathodeSurface).

Two such models are implemented. 
The first model is copied from from geant4.11, which is based on https://ieeexplore.ieee.org/document/9875513
It models the alloy as a thin layer with real refractive index, then calculate reflection probability. However it does not separate the absorption and transmission probability so it cannot model the transmitted photons. Also it cannot handle total internal reflection when n1<n2 (from glass to air/vacuum), so there will be no photon absorption beyond the critical angle.

The second model is based on https://arxiv.org/abs/physics/0408075v1
It models the alloy as a thin layer with real and imaginary refractive indices, then calculate absorption, reflection and transmission probability. It is an extension of the first model which can handle photon transmission and absorption beyond critical angle, but needs the extra imaginary refractive indices to quantity the absorption process. 

Below are the photon absorption, transmission and reflection curves for a specific set of parameters. The incident angle is the angle relative to the glass surface normal instead of the PMT normal. Both models show a relative stable curve before the critical angle (but the values are quite different), and the behavior beyond is wildly different. In principle the second model should be more physical but we will need validations from calibration data.
image
image

Some custom classes are declared to order to use the custom models. `WCSimOpBoundaryProcess`, `WCSimOpticalPhysics` and `WCSimOpticalPhysicsMessenger` classes are modified from the corresponding geant4 classes from geant4.10.05p01 and backward compatible with the official version (geant4.10.01p03) of WCTE.

To use the new models, just change the new tuning parameter `/WCSim/tuning/pmtsurftype` to 1 or 2 in `macros/tuning_parameters.mac`. 
Currently, there are three sets of the predefined photocathode thin film properties that can be chosen through `/WCSim/tuning/cathodepara`. Choice `0` seems to be the parameters being used in SK, while `1` and `2` correspond to the best fit optical parameters for the KCsCb and RbCsCb photocathode measurements from the second model.
But of course we will need external data, e.g. PTF measurement to validate those parameters.

### CODE REVIEW

File: src/WCSimRunAction.cc [hunk 1]
```diff
@@ -181,34 +181,34 @@ void WCSimRunAction::BeginOfRunAction(const G4Run* /*aRun*/)
...
-  geomTree->Branch("PMTradius_ID",&pmt_radius_id,"PMTradius_ID/Double_t");
-  geomTree->Branch("PMTtype_OD",pmt_od_string,"PMTtype_OD[50]/Char_t");
-  geomTree->Branch("PMTradius_OD",&pmt_radius_od,"PMTradius_OD/Double_t");
-  geomTree->Branch("numPMT_ID",&numPMT_id,"numPMT_ID/Int_t");
-  geomTree->Branch("numPMT_OD",&numPMT_od,"numPMT_OD/Int_t");
-  geomTree->Branch("Orientation",&orient,"Orientation/Int_t");
-  geomTree->Branch("Offset_x",&offset_x,"Offset_x/Double_t");
-  geomTree->Branch("Offset_y",&offset_y,"Offset_y/Double_t");
-  geomTree->Branch("Offset_z",&offset_z,"Offset_z/Double_t");
+  geomTree->Branch("GeometryType",geo_type_string,"GeometryType[20]/C");         //example: for std::string data_ : tree->Branch(branchname.c_str(), (void*)data_->c_str(),leafdescription.c_str());
```
Comment (pdeperio) [inline src/WCSimRunAction.cc]: Split to separate PR

File: src/WCSimVisManager.cc [hunk 2]
```diff
@@ -141,7 +141,8 @@ void WCSimVisManager::RegisterGraphicsSystems () {
     G4cout <<
       "\nYou have successfully chosen to use the following graphics systems."
 	 << G4endl;
-    PrintAvailableGraphicsSystems ();
+    //PrintAvailableGraphicsSystems (); //use this version for Geant4.10.1
```
Comment (pdeperio) [inline src/WCSimVisManager.cc]: Split to separate PR (using precompiler and Geant version environment variable if possible)

File: src/WCSimConstructMaterials.cc [hunk 3]
```diff
@@ -1085,6 +1085,13 @@ void WCSimDetectorConstruction::ConstructMaterials()
    G4double EFFICIENCY_glasscath[NUM] =
      { 0.0, 0.0 };
 
+   // Coated surface properties for new photocathode physics
```
Comment (pdeperio) [inline src/WCSimConstructMaterials.cc]: Add macro commands to allow variation/fitting studies later?

**Review by pdeperio — CHANGES_REQUESTED** [review]
add link to documentation and diagrams/plots here

File: src/WCSimConstructMaterials.cc (hunk 3 above)
Comment (xligit) [inline src/WCSimConstructMaterials.cc]: Can you specify what these parameters mean, so it's easy for other people to follow?

File: src/WCSimConstructMaterials.cc [hunk 4]
```diff
@@ -1201,6 +1208,26 @@ void WCSimDetectorConstruction::ConstructMaterials()
    myST2->AddProperty("REFLECTIVITY", PP, REFLECTIVITY_glasscath, NUM);
    myST2->AddProperty("EFFICIENCY", PP, EFFICIENCY_glasscath, NUM);
    //myST2->AddProperty("ABSLENGTH", PP, abslength_paint , NUM);
+
+   // In order to use new photocathode physics
+   G4int pmtsurftype = WCSimTuningParams->GetPMTSurfType(); // Choose one of the two models, see WCSimOpBoundaryProcess for model details
+   if (pmtsurftype==1)
+   {
+      OpGlassCathodeSurface->SetType(G4SurfaceType(x_ray + 11));
```
Comment (xligit) [inline src/WCSimConstructMaterials.cc]: Can you add some comments here?

File: src/WCSimRunAction.cc (hunk 1 above)
Comment (kmtsui) [reply src/WCSimRunAction.cc]: resolved in #10

File: src/WCSimVisManager.cc (hunk 2 above)
Comment (kmtsui) [reply src/WCSimVisManager.cc]: resolved in #10

File: src/WCSimConstructMaterials.cc (hunk 3 above)
Comment (kmtsui) [reply src/WCSimConstructMaterials.cc]: The refractive indices are in general a function of photon energy/wavelength. Need to think of a way to allow variation with a single or few parameters, e.g. using some pre-defined sets and parameteric forms

Comment (pdeperio) [reply src/WCSimConstructMaterials.cc]: @kmtsui i.e. which equation and variable in the paper(s) these correspond to?

Comment (pdeperio) [reply src/WCSimConstructMaterials.cc]: i guess related to Xiaoyue's request below, then are these parameters just assumed constant for all wavelength currently?

do you want to try to implement wavelength dependence and variations in this PR?

Comment (kmtsui) [reply src/WCSimConstructMaterials.cc]: Just implemented some wavelength dependent sets of refractive indices

Comment (kmtsui) [reply src/WCSimConstructMaterials.cc]: These are the refractive indices of the photocathode film. Better to directly refer to`WCSimOpBoundaryProcess` to see how they are used.

**Review by pdeperio — APPROVED** [review]

**Review by laurenanthony2 — APPROVED** [review]

File: include/WCSimOpBoundaryProcess.hh [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+
+        G4int iTE, iTM;
+
+        G4double kCarTolerance;
+
+        size_t idx, idy;
+        G4Physics2DVector* DichroicVector;
+
+        // Used by CoatedDielectricDielectric() and CoatedDielectricDielectric_alt()
+        G4double fCoatedRindex, fCoatedRindexIm, fCoatedThickness;
```
Comment (xligit) [inline include/WCSimOpBoundaryProcess.hh:288]: is it better to change it to fCoatedRindexRe?

File: src/WCSimOpBoundaryProcess.cc [hunk 6]
```diff
@@ -0,0 +1,2058 @@
...
+  }
+
+  Reflectivity_TE = (rTE * conj(rTE)) * (E1_perp * E1_perp) / (E1_perp * E1_perp + E1_parl * E1_parl);
+  Reflectivity_TM = (rTM * conj(rTM)) * (E1_parl * E1_parl) / (E1_perp * E1_perp + E1_parl * E1_parl);
+  Reflectivity = Reflectivity_TE + Reflectivity_TM;
+
+  return real(Reflectivity);
+}
+
+void WCSimOpBoundaryProcess::CoatedDielectricDielectric_alt()
```
Comment (xligit) [inline src/WCSimOpBoundaryProcess.cc]: Am I understanding it correctly that you do photon tracking in glass by detault Geant4 process? Why not include the water-glass boundary as well in this function like in the paper?

**Review by xligit — APPROVED** [review]
